File: go/internal/application/ports.go
// Package application orchestrates the canonical place-order use case.
//
// The package owns the outbound ports: adapters implement them, never the
// other way around. Every fallible port returns Go's error interface; a nil
// error is success and a non-nil error wraps a typed domain failure
// (InvalidOrder, InsufficientStock, PaymentDeclined, PersistenceConflict,
// CompensationFailure, or OrderAlreadyShipped), recoverable with errors.As.
package application

import (
	"context"
	"time"

	"warehouse/internal/domain"
)

// ReservationToken is proof that stock for an order was reserved atomically.
type ReservationToken struct {
	OrderID        domain.OrderID
	IdempotencyKey string
}

// ChargeReceipt is proof that payment was collected for an idempotency key.
type ChargeReceipt struct {
	OrderID        domain.OrderID
	IdempotencyKey string
}

// OrderIDGenerator is the outbound port that mints order identifiers.
// Production adapters may use a CSPRNG; tests inject deterministic fakes.
// The domain never reads randomness itself.
type OrderIDGenerator interface {
	Next() domain.OrderID
}

// InventoryGateway is the outbound port for atomic stock reservation.
// ReserveAll covers every line or none; a shortage wraps InsufficientStock.
// Release failure wraps CompensationFailure.
type InventoryGateway interface {
	ReserveAll(ctx context.Context, orderID domain.OrderID, lines []domain.OrderLine, idempotencyKey string) (ReservationToken, error)
	Release(ctx context.Context, token ReservationToken) error
}

// PaymentProcessor is the outbound port for idempotent payment collection.
// Charge wraps PaymentDeclined when refused. Refund failure wraps
// CompensationFailure. Identical idempotency keys replay the original outcome.
type PaymentProcessor interface {
	Charge(ctx context.Context, order *domain.Order, idempotencyKey string) (ChargeReceipt, error)
	Refund(ctx context.Context, receipt ChargeReceipt) error
}

// OrderRepository is the outbound port that persists and retrieves snapshots.
// Save uses compare-and-set expectedVersion and wraps PersistenceConflict on
// a lost race. Get reports absence with the second return value; an unknown
// identifier never raises. Reads and writes never expose a mutable alias.
type OrderRepository interface {
	Save(ctx context.Context, order *domain.Order, expectedVersion int) (*domain.Order, error)
	Get(ctx context.Context, id domain.OrderID) (*domain.Order, bool)
	GetByIdempotencyKey(ctx context.Context, key string) (fingerprint string, order *domain.Order, ok bool)
	RememberIdempotency(ctx context.Context, key, fingerprint string, order *domain.Order)
}

// Clock provides the current instant. Production code must inject it instead
// of calling time.Now directly, so tests pin time deterministically.
type Clock interface {
	Now() time.Time
}

File: swift/Sources/Warehouse/Application/Ports.swift
/// Proof that stock for an order was reserved atomically.
public struct ReservationToken: Equatable, Sendable {
  public let orderId: OrderId
  public let idempotencyKey: String

  public init(orderId: OrderId, idempotencyKey: String) {
    self.orderId = orderId
    self.idempotencyKey = idempotencyKey
  }
}

/// Proof that payment was collected for an idempotency key.
public struct ChargeReceipt: Equatable, Sendable {
  public let orderId: OrderId
  public let idempotencyKey: String

  public init(orderId: OrderId, idempotencyKey: String) {
    self.orderId = orderId
    self.idempotencyKey = idempotencyKey
  }
}

/// Fingerprint and snapshot for a previous successful command.
public struct IdempotencyRecord: Equatable, Sendable {
  public let fingerprint: String
  public let order: Order

  public init(fingerprint: String, order: Order) {
    self.fingerprint = fingerprint
    self.order = order
  }
}

/// Outbound port that mints deterministic-in-tests order identifiers.
public protocol OrderIdGenerator: Sendable {
  func next() -> OrderId
}

/// Outbound port for atomic stock reservation.
public protocol InventoryGateway: Sendable {
  func reserveAll(
    orderId: OrderId,
    lines: [OrderLine],
    idempotencyKey: String
  ) -> Result<ReservationToken, InsufficientStock>

  func release(token: ReservationToken) -> Result<Void, CompensationFailure>
}

/// Outbound port for idempotent payment collection.
public protocol PaymentProcessor: Sendable {
  func charge(_ order: Order, idempotencyKey: String) -> Result<ChargeReceipt, PaymentDeclined>
  func refund(_ receipt: ChargeReceipt) -> Result<Void, CompensationFailure>
}

/// Outbound port that persists and retrieves immutable snapshots.
public protocol OrderRepository: Sendable {
  func save(_ order: Order, expectedVersion: Int) -> Result<Order, PersistenceConflict>
  func get(_ orderId: OrderId) -> Order?
  func getByIdempotencyKey(_ key: String) -> IdempotencyRecord?
  func rememberIdempotency(key: String, fingerprint: String, order: Order)
}

/// Current instant. Production code must inject this instead of reading the clock.
public protocol Clock: Sendable {
  /// Unix epoch milliseconds. The domain never reads wall-clock time itself.
  func now() -> Int64
}
